### ISSUE TITLE
Reduction in tflite performance between v2.5 and v2.9

### ISSUE THREAD

**Issue by parnham** [tensorflow/tensorflow#56463]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Performance

### Source

source

### Tensorflow Version

tflite 2.5.0-rc3 to 2.9.1

### Custom Code

Yes

### OS Platform and Distribution

Ubuntu 20.04

### Mobile device

n/a

### Python version

n/a

### Bazel version

4.2.1 for earlier tflite versions, 5.1.1 for v2.9.1

### GCC/Compiler version

9.4.0-1ubuntu1~20.04.1

### CUDA/cuDNN version

n/a

### GPU model and memory

NVIDIA RTX 2070

### Current Behaviour?

```shell
We are seeing a decline in the inferencing performance of tflite using the opencl gpu delegate on an x86_64 platform.

The log output below shows the performance of each version we've tested. The classifier, hardware and our software is identical in each case - the only change is the version of tflite. The results are the average time per sample over several thousand inferences.

The classifier model is inception-v1 based.

Since batch processing is still not possible in the tflite gpu delegate then single inference performance is critical.

We're still using 2.5.0-rc3 in production due to this performance issue.
```

### Standalone code to reproduce the issue

```shell
Since the classifier was trained for a client on proprietary data we are unable to post the classifier and the test samples here. 
However the second part of the log shows the results using the tflite benchmark_model tool built for x64 with opencl delegate support. It shows the same increase in average inference time and weirdly the 2.8 and 2.9 versions also segfaulted after the benchmark.
```

### Relevant log output

```shell
## Using our framework
Version    Time to load (ms)	Time/sample (ms)
2.5.0-rc3  215                  1.76
2.5.3      222                  1.83
2.6.5      289                  2.08
2.8.2      2173                 3.35
2.9.1      1727                 3.33

---

## Using benchmark_model
dan@dodeca:[local *]:~/code/external$ ./bin/benchmark_model_v2.5.0-rc3 --use_gpu=true --graph=/home/dan/code/data-eagle/libeagle-data-cig12/usr/share/libeagle/normal/model.tflite
STARTING!
Log parameter values verbosely: [0]
Graph: [/home/dan/code/data-eagle/libeagle-data-cig12/usr/share/libeagle/normal/model.tflite]
Use gpu: [1]
Loaded model /home/dan/code/data-eagle/libeagle-data-cig12/usr/share/libeagle/normal/model.tflite
INFO: Created TensorFlow Lite delegate for GPU.
INFO: Initialized OpenCL-based API.
INFO: Created 1 GPU delegate kernels.
Explicitly applied GPU delegate, and the model graph will be completely executed by the delegate.
The input model file size (MB): 22.452
Initialized session in 154.524ms.
Running benchmark for at least 1 iterations and at least 0.5 seconds but terminate if exceeding 150 seconds.
count=237 first=8216 curr=1821 min=1723 max=8216 avg=2096.54 std=583

Running benchmark for at least 50 iterations and at least 1 seconds but terminate if exceeding 150 seconds.
count=526 first=2277 curr=1732 min=1725 max=2765 avg=1884.21 std=212

Inference timings in us: Init: 154524, First inference: 8216, Warmup (avg): 2096.54, Inference (avg): 1884.21
Note: as the benchmark tool itself affects memory footprint, the following is only APPROXIMATE to the actual memory footprint of the model at runtime. Take the information at your discretion.
Peak memory footprint (MB): init=212.746 overall=212.746

---

dan@dodeca:[local *]:~/code/external$ ./bin/benchmark_model_v2.5.3 --use_gpu=true --graph=/home/dan/code/data-eagle/libeagle-data-cig12/usr/share/libeagle/normal/model.tflite
STARTING!
Log parameter values verbosely: [0]
Graph: [/home/dan/code/data-eagle/libeagle-data-cig12/usr/share/libeagle/normal/model.tflite]
Use gpu: [1]
Loaded model /home/dan/code/data-eagle/libeagle-data-cig12/usr/share/libeagle/normal/model.tflite
INFO: Created TensorFlow Lite delegate for GPU.
INFO: Initialized OpenCL-based API.
INFO: Created 1 GPU delegate kernels.
Explicitly applied GPU delegate, and the model graph will be completely executed by the delegate.
The input model file size (MB): 22.452
Initialized session in 168.928ms.
Running benchmark for at least 1 iterations and at least 0.5 seconds but terminate if exceeding 150 seconds.
count=233 first=10964 curr=1728 min=1728 max=10964 avg=2127.98 std=719

Running benchmark for at least 50 iterations and at least 1 seconds but terminate if exceeding 150 seconds.
count=523 first=1749 curr=1735 min=1727 max=2719 avg=1896.62 std=214

Inference timings in us: Init: 168928, First inference: 10964, Warmup (avg): 2127.98, Inference (avg): 1896.62
Note: as the benchmark tool itself affects memory footprint, the following is only APPROXIMATE to the actual memory footprint of the model at runtime. Take the information at your discretion.
Peak memory footprint (MB): init=211.449 overall=211.449

---

dan@dodeca:[local *]:~/code/external$ ./bin/benchmark_model_v2.6.5 --use_gpu=true --graph=/home/dan/code/data-eagle/libeagle-data-cig12/usr/share/libeagle/normal/model.tflite
STARTING!
Log parameter values verbosely: [0]
Graph: [/home/dan/code/data-eagle/libeagle-data-cig12/usr/share/libeagle/normal/model.tflite]
Use gpu: [1]
Loaded model /home/dan/code/data-eagle/libeagle-data-cig12/usr/share/libeagle/normal/model.tflite
INFO: Created TensorFlow Lite delegate for GPU.
GPU delegate created.
Going to apply 1 delegates one after another.
INFO: Initialized OpenCL-based API.
INFO: Created 1 GPU delegate kernels.
Explicitly applied GPU delegate, and the model graph will be completely executed by the delegate.
The input model file size (MB): 22.452
Initialized session in 163.602ms.
Running benchmark for at least 1 iterations and at least 0.5 seconds but terminate if exceeding 150 seconds.
count=228 first=8352 curr=1788 min=1780 max=8352 avg=2106.11 std=585

Running benchmark for at least 50 iterations and at least 1 seconds but terminate if exceeding 150 seconds.
count=500 first=1799 curr=1802 min=1777 max=7597 avg=1911.31 std=309

Inference timings in us: Init: 163602, First inference: 8352, Warmup (avg): 2106.11, Inference (avg): 1911.31
Note: as the benchmark tool itself affects memory footprint, the following is only APPROXIMATE to the actual memory footprint of the model at runtime. Take the information at your discretion.
Peak memory footprint (MB): init=211.082 overall=211.082

---

dan@dodeca:[local *]:~/code/external$ ./bin/benchmark_model_v2.8.2 --use_gpu=true --graph=/home/dan/code/data-eagle/libeagle-data-cig12/usr/share/libeagle/normal/model.tflite
STARTING!
Log parameter values verbosely: [0]
Graph: [/home/dan/code/data-eagle/libeagle-data-cig12/usr/share/libeagle/normal/model.tflite]
Use gpu: [1]
Loaded model /home/dan/code/data-eagle/libeagle-data-cig12/usr/share/libeagle/normal/model.tflite
INFO: Created TensorFlow Lite delegate for GPU.
GPU delegate created.
INFO: Initialized OpenCL-based API.
INFO: Created 1 GPU delegate kernels.
Explicitly applied GPU delegate, and the model graph will be completely executed by the delegate.
The input model file size (MB): 22.452
Initialized session in 179.387ms.
Running benchmark for at least 1 iterations and at least 0.5 seconds but terminate if exceeding 150 seconds.
count=132 first=14658 curr=3492 min=3165 max=14658 avg=3706.23 std=1029

Running benchmark for at least 50 iterations and at least 1 seconds but terminate if exceeding 150 seconds.
count=280 first=3877 curr=3433 min=3149 max=8949 avg=3494.44 std=450

Inference timings in us: Init: 179387, First inference: 14658, Warmup (avg): 3706.23, Inference (avg): 3494.44
Note: as the benchmark tool itself affects memory footprint, the following is only APPROXIMATE to the actual memory footprint of the model at runtime. Take the information at your discretion.
Memory footprint delta from the start of the tool (MB): init=211.52 overall=211.52
Segmentation fault (core dumped)

---

dan@dodeca:[local *]:~/code/external$ ./bin/benchmark_model_v2.9.1 --use_gpu=true --graph=/home/dan/code/data-eagle/libeagle-data-cig12/usr/share/libeagle/normal/model.tflite
STARTING!
Log parameter values verbosely: [0]
Graph: [/home/dan/code/data-eagle/libeagle-data-cig12/usr/share/libeagle/normal/model.tflite]
Use gpu: [1]
Loaded model /home/dan/code/data-eagle/libeagle-data-cig12/usr/share/libeagle/normal/model.tflite
INFO: Created TensorFlow Lite delegate for GPU.
GPU delegate created.
INFO: Initialized OpenCL-based API.
INFO: Created 1 GPU delegate kernels.
Explicitly applied GPU delegate, and the model graph will be completely executed by the delegate.
The input model file size (MB): 22.452
Initialized session in 154.443ms.
Running benchmark for at least 1 iterations and at least 0.5 seconds but terminate if exceeding 150 seconds.
count=127 first=13799 curr=3222 min=3179 max=13799 avg=3866.95 std=1039

Running benchmark for at least 50 iterations and at least 1 seconds but terminate if exceeding 150 seconds.
count=279 first=3525 curr=3185 min=3174 max=4197 avg=3510.01 std=271

Inference timings in us: Init: 154443, First inference: 13799, Warmup (avg): 3866.95, Inference (avg): 3510.01
Note: as the benchmark tool itself affects memory footprint, the following is only APPROXIMATE to the actual memory footprint of the model at runtime. Take the information at your discretion.
Memory footprint delta from the start of the tool (MB): init=211.57 overall=211.57
Segmentation fault (core dumped)
```
</details>

**Comment by mohantym** [comment]
Hi @parnham ! Thanks for reporting this issue. @sachinprasadhs ! Could you look at this issue. Thank you!

**Comment by parnham** [comment]
I also have a question related to the GPU performance. 

If I run two instances of our benchmark simultaneously each runs at the same speed as a single one (utilising ~66% of the GPU instead of ~33%). Therefore it stands to reason that running two threads within a process, each with their own model, interpreter and delegate, should provide an inferencing speed increase by classifying two samples at a time. However this is not the case - is there something in the GPU delegate which is preventing concurrent OpenCL pipelines within the same process?

**Comment by korkland** [comment]
Can you please elaborate on the building process with opencl?

**Comment by roserg** [comment]
Hi,
If you want better performance statistics, you can try to use this script:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/delegates/gpu/cl/testing/run_performance_profiling.sh
./run_performance_profiling.sh -m path_to_the_model -d HOST

**Comment by parnham** [comment]
> Hi, If you want better performance statistics, you can try to use this script: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/delegates/gpu/cl/testing/run_performance_profiling.sh ./run_performance_profiling.sh -m path_to_the_model -d HOST

Have attempted to run this script for both the v2.5.0-rc3 and v2.9.1 tags and it fails as follows
```
Starting local Bazel server and connecting to it...
INFO: Options provided by the client:
  Inherited 'common' options: --isatty=1 --terminal_columns=259
INFO: Reading rc options for 'build' from /home/dan/downloads/tensorflow-testing/tensorflow-2.9.1/.bazelrc:
  Inherited 'common' options: --experimental_repo_remote_exec
INFO: Reading rc options for 'build' from /home/dan/downloads/tensorflow-testing/tensorflow-2.9.1/.bazelrc:
  'build' options: --define framework_shared_object=true --define=use_fast_cpp_protos=true --define=allow_oversize_protos=true --spawn_strategy=standalone -c opt --announce_rc --define=grpc_no_ares=true --noincompatible_remove_legacy_whole_archive --enable_platform_specific_config --define=with_xla_support=true --config=short_logs --config=v2 --define=no_aws_support=true --define=no_hdfs_support=true --experimental_cc_shared_library --deleted_packages=tensorflow/compiler/mlir/tfrt,tensorflow/compiler/mlir/tfrt/benchmarks,tensorflow/compiler/mlir/tfrt/jit/python_binding,tensorflow/compiler/mlir/tfrt/jit/transforms,tensorflow/compiler/mlir/tfrt/python_tests,tensorflow/compiler/mlir/tfrt/tests,tensorflow/compiler/mlir/tfrt/tests/ir,tensorflow/compiler/mlir/tfrt/tests/analysis,tensorflow/compiler/mlir/tfrt/tests/jit,tensorflow/compiler/mlir/tfrt/tests/lhlo_to_tfrt,tensorflow/compiler/mlir/tfrt/tests/tf_to_corert,tensorflow/compiler/mlir/tfrt/tests/tf_to_tfrt_data,tensorflow/compiler/mlir/tfrt/tests/saved_model,tensorflow/compiler/mlir/tfrt/transforms/lhlo_gpu_to_tfrt_gpu,tensorflow/core/runtime_fallback,tensorflow/core/runtime_fallback/conversion,tensorflow/core/runtime_fallback/kernel,tensorflow/core/runtime_fallback/opdefs,tensorflow/core/runtime_fallback/runtime,tensorflow/core/runtime_fallback/util,tensorflow/core/tfrt/common,tensorflow/core/tfrt/eager,tensorflow/core/tfrt/eager/backends/cpu,tensorflow/core/tfrt/eager/backends/gpu,tensorflow/core/tfrt/eager/core_runtime,tensorflow/core/tfrt/eager/cpp_tests/core_runtime,tensorflow/core/tfrt/gpu,tensorflow/core/tfrt/run_handler_thread_pool,tensorflow/core/tfrt/runtime,tensorflow/core/tfrt/saved_model,tensorflow/core/tfrt/graph_executor,tensorflow/core/tfrt/saved_model/tests,tensorflow/core/tfrt/tpu,tensorflow/core/tfrt/utils
INFO: Found applicable config definition build:short_logs in file /home/dan/downloads/tensorflow-testing/tensorflow-2.9.1/.bazelrc: --output_filter=DONT_MATCH_ANYTHING
INFO: Found applicable config definition build:v2 in file /home/dan/downloads/tensorflow-testing/tensorflow-2.9.1/.bazelrc: --define=tf_api_version=2 --action_env=TF2_BEHAVIOR=1
INFO: Found applicable config definition build:linux in file /home/dan/downloads/tensorflow-testing/tensorflow-2.9.1/.bazelrc: --copt=-w --host_copt=-w --define=PREFIX=/usr --define=LIBDIR=$(PREFIX)/lib --define=INCLUDEDIR=$(PREFIX)/include --define=PROTOBUF_INCLUDE_PATH=$(PREFIX)/include --cxxopt=-std=c++14 --host_cxxopt=-std=c++14 --config=dynamic_kernels --distinct_host_configuration=false --experimental_guard_against_concurrent_changes
INFO: Found applicable config definition build:dynamic_kernels in file /home/dan/downloads/tensorflow-testing/tensorflow-2.9.1/.bazelrc: --define=dynamic_loaded_kernels=true --copt=-DAUTOLOAD_DYNAMIC_KERNELS
ERROR: Skipping '//.:performance_profiling': invalid target format '//.:performance_profiling': invalid package name '.': package name component contains only '.' characters
WARNING: Target pattern parsing failed.
ERROR: invalid target format '//.:performance_profiling': invalid package name '.': package name component contains only '.' characters
INFO: Elapsed time: 18.457s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded)
chmod: cannot access 'bazel-bin/./performance_profiling': No such file or directory
./run_performance_profiling.sh: line 80: ./bazel-bin/./performance_profiling: No such file or directory
```
What am I doing wrong?

**Comment by roserg** [comment]
What is the result of this command for you:
blaze build -c opt third_party/tensorflow/lite/delegates/gpu/cl/testing:performance_profiling

**Comment by parnham** [comment]
> What is the result of this command for you: bazel build -c opt third_party/tensorflow/lite/delegates/gpu/cl/testing:performance_profiling

It fails with 
```
dan@dodeca:[(v2.9.1)]:~/downloads/tensorflow-testing/tensorflow-2.9.1$ bazel build -c opt third_party/tensorflow/lite/delegates/gpu/cl/testing:performance_profiling
Starting local Bazel server and connecting to it...
INFO: Options provided by the client:
  Inherited 'common' options: --isatty=1 --terminal_columns=259
INFO: Reading rc options for 'build' from /home/dan/downloads/tensorflow-testing/tensorflow-2.9.1/.bazelrc:
  Inherited 'common' options: --experimental_repo_remote_exec
INFO: Reading rc options for 'build' from /home/dan/downloads/tensorflow-testing/tensorflow-2.9.1/.bazelrc:
  'build' options: --define framework_shared_object=true --define=use_fast_cpp_protos=true --define=allow_oversize_protos=true --spawn_strategy=standalone -c opt --announce_rc --define=grpc_no_ares=true --noincompatible_remove_legacy_whole_archive --enable_platform_specific_config --define=with_xla_support=true --config=short_logs --config=v2 --define=no_aws_support=true --define=no_hdfs_support=true --experimental_cc_shared_library --deleted_packages=tensorflow/compiler/mlir/tfrt,tensorflow/compiler/mlir/tfrt/benchmarks,tensorflow/compiler/mlir/tfrt/jit/python_binding,tensorflow/compiler/mlir/tfrt/jit/transforms,tensorflow/compiler/mlir/tfrt/python_tests,tensorflow/compiler/mlir/tfrt/tests,tensorflow/compiler/mlir/tfrt/tests/ir,tensorflow/compiler/mlir/tfrt/tests/analysis,tensorflow/compiler/mlir/tfrt/tests/jit,tensorflow/compiler/mlir/tfrt/tests/lhlo_to_tfrt,tensorflow/compiler/mlir/tfrt/tests/tf_to_corert,tensorflow/compiler/mlir/tfrt/tests/tf_to_tfrt_data,tensorflow/compiler/mlir/tfrt/tests/saved_model,tensorflow/compiler/mlir/tfrt/transforms/lhlo_gpu_to_tfrt_gpu,tensorflow/core/runtime_fallback,tensorflow/core/runtime_fallback/conversion,tensorflow/core/runtime_fallback/kernel,tensorflow/core/runtime_fallback/opdefs,tensorflow/core/runtime_fallback/runtime,tensorflow/core/runtime_fallback/util,tensorflow/core/tfrt/common,tensorflow/core/tfrt/eager,tensorflow/core/tfrt/eager/backends/cpu,tensorflow/core/tfrt/eager/backends/gpu,tensorflow/core/tfrt/eager/core_runtime,tensorflow/core/tfrt/eager/cpp_tests/core_runtime,tensorflow/core/tfrt/gpu,tensorflow/core/tfrt/run_handler_thread_pool,tensorflow/core/tfrt/runtime,tensorflow/core/tfrt/saved_model,tensorflow/core/tfrt/graph_executor,tensorflow/core/tfrt/saved_model/tests,tensorflow/core/tfrt/tpu,tensorflow/core/tfrt/utils
INFO: Found applicable config definition build:short_logs in file /home/dan/downloads/tensorflow-testing/tensorflow-2.9.1/.bazelrc: --output_filter=DONT_MATCH_ANYTHING
INFO: Found applicable config definition build:v2 in file /home/dan/downloads/tensorflow-testing/tensorflow-2.9.1/.bazelrc: --define=tf_api_version=2 --action_env=TF2_BEHAVIOR=1
INFO: Found applicable config definition build:linux in file /home/dan/downloads/tensorflow-testing/tensorflow-2.9.1/.bazelrc: --copt=-w --host_copt=-w --define=PREFIX=/usr --define=LIBDIR=$(PREFIX)/lib --define=INCLUDEDIR=$(PREFIX)/include --define=PROTOBUF_INCLUDE_PATH=$(PREFIX)/include --cxxopt=-std=c++14 --host_cxxopt=-std=c++14 --config=dynamic_kernels --distinct_host_configuration=false --experimental_guard_against_concurrent_changes
INFO: Found applicable config definition build:dynamic_kernels in file /home/dan/downloads/tensorflow-testing/tensorflow-2.9.1/.bazelrc: --define=dynamic_loaded_kernels=true --copt=-DAUTOLOAD_DYNAMIC_KERNELS
ERROR: Skipping 'third_party/tensorflow/lite/delegates/gpu/cl/testing:performance_profiling': no such package 'third_party/tensorflow/lite/delegates/gpu/cl/testing': BUILD file not found in any of the following directories. Add a BUILD file to a directory to mark it as a package.
 - /home/dan/downloads/tensorflow-testing/tensorflow-2.9.1/third_party/tensorflow/lite/delegates/gpu/cl/testing
WARNING: Target pattern parsing failed.
ERROR: no such package 'third_party/tensorflow/lite/delegates/gpu/cl/testing': BUILD file not found in any of the following directories. Add a BUILD file to a directory to mark it as a package.
 - /home/dan/downloads/tensorflow-testing/tensorflow-2.9.1/third_party/tensorflow/lite/delegates/gpu/cl/testing
INFO: Elapsed time: 18.309s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded)
```
But that path doesn't exist, so if I drop the `third_party` at the beginning then it builds.

**Comment by parnham** [comment]
Having built the tool using the correct path I was able to produce a comparison of 
[v2.5.0-rc3.log](https://github.com/tensorflow/tensorflow/files/9121728/profile-v2.5.0-rc3.log) and 
[v2.9.1.log](https://github.com/tensorflow/tensorflow/files/9121734/profile-v2.9.1.log)

I've also put it in a spreadsheet so that it's easier to compare the timings:
https://docs.google.com/spreadsheets/d/1xmcY5TMXFH3RvSE91ZPuvLe2OZxqfGNbw5h7SR0ibug/edit?usp=sharing

It seems that later tflite implementations use winograd in place of certain 2D convolutions, but according to the profiles both have similar ideal total times (with older tflite being slightly quicker). 

But then during inferencing v2.5.0-rc3 beats the ideal time considerably whereas v2.9.1 is much slower than the ideal time.

**Comment by roserg** [comment]
Hi,
Thank you for your information
From my side I can see that 'usual inference time' higher than ideal total time until this CL: https://github.com/tensorflow/tensorflow/commit/14dd3a808ded246021a110ee7cf8b174214ed6c2
and much smaller after.
I don't know if this change included in v2.9.1
I hope this CL will fix performance for you.

"It seems that later tflite implementations use winograd in place of certain 2D convolutions, but according to the profiles both have similar ideal total times (with older tflite being slightly quicker)."
Nope, both of them use Winograd, but in version 2.5 this profiling tool was not so detailed and used convolution name for Winograd transformations.

**Comment by roserg** [comment]
> I also have a question related to the GPU performance.
> 
> If I run two instances of our benchmark simultaneously each runs at the same speed as a single one (utilising ~66% of the GPU instead of ~33%). Therefore it stands to reason that running two threads within a process, each with their own model, interpreter and delegate, should provide an inferencing speed increase by classifying two samples at a time. However this is not the case - is there something in the GPU delegate which is preventing concurrent OpenCL pipelines within the same process?

I don't know why it works as you described and can not help with this question.

**Comment by parnham** [comment]
> Hi, Thank you for your information From my side I can see that 'usual inference time' higher than ideal total time until this CL: [14dd3a8](https://github.com/tensorflow/tensorflow/commit/14dd3a808ded246021a110ee7cf8b174214ed6c2) and much smaller after. I don't know if this change included in v2.9.1 I hope this CL will fix performance for you.
> 
> "It seems that later tflite implementations use winograd in place of certain 2D convolutions, but according to the profiles both have similar ideal total times (with older tflite being slightly quicker)." Nope, both of them use Winograd, but in version 2.5 this profiling tool was not so detailed and used convolution name for Winograd transformations.

You're right - your patch in April did fix this issue but unfortunately did not make it in to the v2.9.1 release! I've updated the spreadsheet with performance results from the master branch and it has indeed returned to the inference speeds we used to see. We'll just have to wait for the next tagged release and then we can update our dependencies.

Many thanks.
